cluster-autoscaler/cloudprovider/azure/azure_cache.go

```diff
@@ -21,96 +21,234 @@ import (
+		virtualMachines:      make(map[string][]compute.VirtualMachine),
+		registeredNodeGroups: make([]cloudprovider.NodeGroup, 0),
+		instanceToNodeGroup:  make(map[azureRef]cloudprovider.NodeGroup),
+		unownedInstances:     make(map[azureRef]bool),
 	}
 
-	cache.mutex.Lock()
-	defer cache.mutex.Unlock()
 	if err := cache.regenerate(); err != nil {
-		klog.Errorf("Error while regenerating Asg cache: %v", err)
+		klog.Errorf("Error while regenerating Azure cache: %v", err)
 	}
 
 	return cache, nil
 }
 
+func (m *azureCache) getVirtualMachines() map[string][]compute.VirtualMachine {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.virtualMachines
+}
+
+func (m *azureCache) getScaleSets() map[string]compute.VirtualMachineScaleSet {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.scaleSets
+}
+
+// Cleanup closes the channel to signal the go routine to stop that is handling the cache
+func (m *azureCache) Cleanup() {
+	close(m.interrupt)
+}
+
+func (m *azureCache) regenerate() error {
+	err := m.fetchAzureResources()
+	if err != nil {
+		return err
+	}
+
+	// Regenerate instance to node groups mapping.
+	newInstanceToNodeGroupCache := make(map[azureRef]cloudprovider.NodeGroup)
+	for _, ng := range m.registeredNodeGroups {
+		klog.V(2).Infof("regenerate: finding nodes for node group %+v", ng)
+		instances, err := ng.Nodes()
+		if err != nil {
+			return err
+		}
+		klog.V(4).Infof("regenerate: found nodes for node group %v: %+v", ng, instances)
+
+		for _, instance := range instances {
+			ref := azureRef{Name: instance.Id}
+			newInstanceToNodeGroupCache[ref] = ng
+		}
+	}
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	m.instanceToNodeGroup = newInstanceToNodeGroupCache
+
+	// Reset unowned instances cache.
+	m.unownedInstances = make(map[azureRef]bool)
+
+	return nil
+}
+
+func (m *azureCache) fetchAzureResources() error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	// List all VMs in the RG.
+	vmResult, err := m.fetchVirtualMachines()
+	if err == nil {
+		m.virtualMachines = vmResult
+	} else {
+		return err
+	}
+
+	// List all VMSS in the RG.
+	vmssResult, err := m.fetchScaleSets()
+	if err == nil {
+		m.scaleSets = vmssResult
+	} else {
+		return err
+	}
+	return nil
+}
+
+// fetchVirtualMachines returns the updated list of virtual machines in the config resource group using the Azure API.
+func (m *azureCache) fetchVirtualMachines() (map[string][]compute.VirtualMachine, error) {
+	ctx, cancel := getContextWithCancel()
+	defer cancel()
+
+	result, err := m.azClient.virtualMachinesClient.List(ctx, m.resourceGroup)
+	if err != nil {
+		klog.Errorf("VirtualMachinesClient.List in resource group \"%s\" failed: %v", m.resourceGroup, err)
+		return nil, err.Error()
+	}
+
+	instances := make(map[string][]compute.VirtualMachine)
+	for _, instance := range result {
+		if instance.Tags == nil {
+			continue
+		}
+
+		tags := instance.Tags
+		vmPoolName := tags["poolName"]
```

Comment:
Doesn't this makes an assumption that every "nodegroup" will have this label? Is this true? For people not using AKS-Engine.

Also, this might be part of the "WIP" so ignore if it doesn't apply but doesn't this mean that you're making extra calls here? `Nodes()` in azure_scale_set already lists the VMs in a scale set.

Reply:
> Doesn't this make an assumption that every "nodegroup" will have this label? Is this true? For people not using AKS-Engine.

It does but that's not new behavior, the current code already assumes that https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go#L387

Reply:
> Also, this might be part of the "WIP" so ignore if it doesn't apply but doesn't this mean that you're making extra calls here? Nodes() in azure_scale_set already lists the VMs in a scale set

Are you asking about https://github.com/kubernetes/autoscaler/pull/3717/files#diff-63481ca096e322d8f48e57a4b21089a30481cc3798e3dc123a1b4ba0938ceb1dR103 ? `fetchVirtualMachines` lists the VMs, not the VMSS instances. The separate call to Nodes() above lists the VMSS instances, using the separate VMSS VM cache.

Reply:
gotcha on the first. For the second, unless I'm confused with the changes, `fetchAzureResources` would be called regardless of the operational mode right? So for scale sets, agentpool and aks that would be called. It makes the call to fetch virtual machines in a resource group - from the scale set mode point of view - the result of that call will never be utilized right?

Reply:
Right, this does mean 1 call to list the scale sets in the RG and 1 call to list the VMs in the RG would happen per minute. Part of the reasoning was to truly allow mixed mode clusters in the future (ref https://github.com/kubernetes-sigs/cloud-provider-azure/issues/338#issuecomment-639254680). Right now, the code skips VMs for vmType "vmss" and vice versa. There were also multiple scale set List() calls happening so this is an attempt to centralize everything in one place. Agree that VM list won't be used in a scenario with scale sets only though (since VMs don't support auto-discovery) so it might make sense to only keep the scale sets in the central cache and move VMs back to the VM cache to avoid those extra calls (similar to how VMSS VMs has its own). We would save 60 calls per minute when no VM node groups are present (regardless of number of scale sets) which is virtually nothing compared to 1 VMSS VM List per scale set when there are large numbers of scale sets.

Reply:
LIST VM is not necessary for VMSS-only clusters. Let's only list VMSS when using "vmss" mode.